Model group

wfbraga opened the pull request:
Creating model Groups and it's tests
task: https://trello.com/c/7dxQNU4N/4-criar-modelo-group

wfbraga committed:
Fist commit

wfbraga committed:
Adding gems for development
  gem 'rspec', '~> 3.5'
  gem 'rubocop', '~> 1.29', '>= 1.29.1'
  gem 'solargraph', '~> 0.44.3'

wfbraga committed:
Creating model User
trello task: https://trello.com/c/bBVAh18C/3-criar-modelo-user

It was add a index to colum email and assured its uniquenes in the BD.
In the code uniquenes is  not case sensitive.

wfbraga committed:
User model tests
Task: https://trello.com/c/bBVAh18C/3-criar-modelo-user

wfbraga committed:
Creating Group model. Setting its realation on User model and creating tests

wfbraga committed:
mergin with main

wfbraga committed:
Merge branch 'model-goup' into model-group

brunoao86 commented on app/models/group.rb:
```diff
@@ -1,2 +1,5 @@
 class Group < ApplicationRecord
+  belongs_to :user, foreign_key: 'owner_id'
```
Hey @wfbraga,

I think it would be better to call this relation by `owner`, to get the group's owner with `group.owner` instead of `group.user`.

Specially because we are going to create `group.users` in the future so it can be confusing.

What do you think?

wfbraga replied:
I agree.
I'll make the changes from
`belongs_to :user`
to
`belongs_to :owner`

brunoao86 commented on app/models/user.rb:
```diff
@@ -1,4 +1,5 @@
 class User < ApplicationRecord
+  has_many :groups, class_name: 'group', foreign_key: 'owner_id'
```
I think we have to find a better name here too to represent the groups that this user is owner, because we are going to have a relation with the same name in the future.

We could have something like:

```ruby
user.own_groups # the groups that the user is owner

user.groups # groups that the user is part of
```

What do you think?

wfbraga replied:
How about...
`user.owned_groups`
It translates to 'grupos propios'

wfbraga committed:
Changes to reflect sugestion on pulrequest as per links bellow

Changing the relations belongs_to and has_many on Group and User to a more intuitive name
Group belongs to owner
User has many owned_groups

comment 1: https://github.com/wfbraga/finco/pull/2#discussion_r880488140
comment 2: https://github.com/wfbraga/finco/pull/2#discussion_r880490188

brunoao86 commented on app/models/group.rb:
```diff
@@ -1,2 +1,5 @@
 class Group < ApplicationRecord
+  belongs_to :owner, class_name: 'User', foreign_key: 'owner_id'
+
+  validates :name, :owner_id, presence: true
```
Hey @wfbraga,

I think you can remove the `owner_id` from this validation because the `belongs_to` relation has an [:optional](https://guides.rubyonrails.org/v6.0/association_basics.html#optional) property with `false` as default value, which means that the presence of `owner` is mandatory by default.

```suggestion
  validates :name, presence: true
```

You can check this locally by creating a group without an owner, and you should see duplicated errors:

```ruby
2.7.1 :015 > group.owner
 => nil 
2.7.1 :016 > group.valid?
 => false 
2.7.1 :017 > group.errors.full_messages
 => ["Owner must exist", "Owner can't be blank"]
```

If you apply the suggested change then we are going to have this result:

```ruby
2.7.1 :003 > group.owner = nil
2.7.1 :004 > group.valid?
 => false 
2.7.1 :005 > group.errors.full_messages
 => ["Owner must exist"]
```

wfbraga replied:
I've have tested your suggestion and it worked as you shown.
Also, reading the code it seamed out of place.
I've decided commit the changes direct from you suggestion.

brunoao86 commented on spec/models/group_spec.rb:
```diff
@@ -1,5 +1,38 @@
+    it 'Owner_id cannot be nil or empty' do
+      group.owner_id = nil
+      group.name = 'Test Growp'
+      expect(group).to be_invalid
```
We are going to have few discussions about testing but here it'd be better to check what is wrong with this object to avoid false positive results.

My suggestion is something like:

```suggestion
      expect(group).to be_invalid
      expect(group.errors.full_messages).to contain_exactly('Owner must exist')
```

This way we guarantee that the object is invalid but also why it is invalid.

If you apply this change before[ my previous comment](https://github.com/wfbraga/finco/pull/2/files#r885595026) you can see that you would catch that error duplication beforehand:

```ruby
  1) Group When creating a invalid group Owner_id cannot be nil or empty
     Failure/Error: expect(group.errors.full_messages).to contain_exactly('Owner must exist')
     
       expected collection contained:  ["Owner must exist"]
       actual collection contained:    ["Owner can't be blank", "Owner must exist"]
       the extra elements were:        ["Owner can't be blank"]
```
